File: 0001-0500/0448. Find All Numbers Disappeared in an Array.java
// Easy
// Array
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/find-all-numbers-disappeared-in-an-array/

import java.util.*;

class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        int n = nums.length;
        int[] cnt = new int[n + 1]; // 1-index arr
        for (int x : nums) {
            cnt[x]++;
        }

        List<Integer> res = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            if (cnt[i] == 0)
                res.add(i);
        }
        return res;
    }
}

File: 0001-0500/0448. Find All Numbers Disappeared in an Array.py
# Time:O(n), Space:O(1)

from typing import List

class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums)

        for num in nums:
            index = abs(num) - 1
            if nums[index] > 0:
                nums[index] = -nums[index]

        res = []
        for i in range(n):
            if nums[i] > 0:
                res.append(i + 1)

        return res
